dan_layer/core/src/workers/hotstuff_waiter.rs

```diff
@@ -607,14 +614,28 @@ where
     }
 
     async fn pacemaker_trigger_foreign_leader_fail(
-        &self,
+        &mut self,
         epoch: Epoch,
         payload_id: PayloadId,
         shard_id: ShardId,
     ) -> Result<(), HotStuffError> {
-        // TODO: We should check if this is triggered second time. If that's the case, then foreign committee didn't
-        // responded to our first recovery request. So they probably can't reach consensus and this transaction should
-        // be thrown away.
+        // Check if this is the second time we are triggering a foreign leader request
+        if self
+            .foreign_leader_failures
+            .contains(&PacemakerEvents::ForeignCommittee(epoch, shard_id, payload_id))
+        {
+            // This is the second time we are triggering a foreign leader request, for the same triplet
+            // (epoch, shard_id, payload_id). This means that they probably can't reach consensus, so we throw
+            // away this transaction
+            self.shard_store.with_write_tx(|tx| {
```

Comment:
Doing this on the second foreign leader failure seems arbitrary. In other rejections, we end up voting REJECT once the proposals come through. In this case, we assume (because likely at this stage we've voted already) that other local replicas have come to the same conclusion, which may not be the case because e.g our network is bad - so we must still allow for enough other replicas to vote ACCEPT because they received the proposals (I think anyway). 

This is a tricky problem that probably is addressed in the HotStuff/Cerberus papers. Perhaps let's leave this one out for now.

Reply:
If I understand it correctly, the logic is that if I am not receiving a foreign leader message the pacemaker will trigger after delta time. In that case, I request `f + 1` nodes from that committee the latest QC they have stored, while I start a pacemaker timer for this. If it times out, it means that neither the leader, neither the requested `f + 1` nodes have answered, in which case, they would not be able to progress local consensus (transaction malformed, etc). In this case, we should remove the local substates we have stored, because we are basically rejecting the transaction. @Cifko, do you agree with this ?

Reply:
The substate is not written at this state, is it? In hotstuff, the substate (command) is executed once we are at decide phase. So there is no rollbacking DB. I don't know how the REJECT voting works, but I assume similar to ACCEPT. So there should never be a case when the pacemaker trigger twice, right? Because one of the nodes should send the REJECT. You should double check the code if the reject is send as a proposal to all committees. If so we can be sure that one of the good replicas will send the REJECT. And it that case you can leave it as is (starting the pacemaker again). But make sure the reject is send back as a response.

Reply:
`Because one of the nodes should send the REJECT. `

Here you mean, a local node for the foreign committee ? So, if the consensus cannot be reached at that foreign committee (malformed transaction, etc), at least one good replica will send REJECT to all other replicas (for each involved shard), including the local node we started with. Is this correct ?

Reply:
I think this is already implemented, because the rejection is in the QC, so when you ask for missing proposal. You will get it from at least one honest replica. And the replica will send you the QC with the rejection on the tx_leader/rx_leader channel. So it should be handled in the `on_receive_proposal` function.